Create exposure_ecosystem_matching.py

sarah-hlsn opened the pull request:
Code to create population exposure dataset and merge with coastal ecosystem data

sarah-hlsn committed:
Create exposure_ecosystem_matching.py

chahank commented on 202305_coastal_ecosystems_TC/impact_calculation/exposure_ecosystem_matching.py:
```diff
@@ -0,0 +1,78 @@
+import geopandas as gpd
```
You should probably add a header with the license (see other CLIMADA files) and you can add yourself as author.

chahank commented on 202305_coastal_ecosystems_TC/impact_calculation/exposure_ecosystem_matching.py:
```diff
@@ -0,0 +1,78 @@
+    for pop_year in pop_years:
+        for hab_year in hab_years:
+            # Load data
+            rhab_path = f'/cluster/work/climate/shuelsen/manuscript/data/rhab/basin_clips/rhab_{hab_year}_{basin}.shp'
+            pop_path = f'/cluster/work/climate/shuelsen/manuscript/data/worldpop/wp_{pop_year}_{basin}.shp'
```
These paths are not useful for an external user. Please make relative paths or define default variables where users can define their own paths.

chahank commented on 202305_coastal_ecosystems_TC/impact_calculation/exposure_ecosystem_matching.py:
```diff
@@ -0,0 +1,78 @@
+            join_filter = join_filter.drop_duplicates(subset='exp_index', keep='first')
+
+            # save data
+            join_filter.to_csv(f'{output}rhab_exposure/per_basin/exp_rhab_{hab_year}_wp_{pop_year}_{basin}.csv')
```
c.f. comment above.

chahank commented on 202305_coastal_ecosystems_TC/impact_calculation/exposure_ecosystem_matching.py:
```diff
@@ -0,0 +1,78 @@
+for pop_year in pop_years:
+    for hab_year in hab_years:
+        # load files
+        AP = pd.read_csv(f'{output}rhab_exposure/per_basin/exp_rhab_{hab_year}_wp_{pop_year}_AP.csv')
+        WP = pd.read_csv(f'{output}rhab_exposure/per_basin/exp_rhab_{hab_year}_wp_{pop_year}_WP.csv')
+        IO = pd.read_csv(f'{output}rhab_exposure/per_basin/exp_rhab_{hab_year}_wp_{pop_year}_IO.csv')
+        SH = pd.read_csv(f'{output}rhab_exposure/per_basin/exp_rhab_{hab_year}_wp_{pop_year}_SH.csv')
```
c.f. comment above.

chahank commented on 202305_coastal_ecosystems_TC/impact_calculation/exposure_ecosystem_matching.py:
```diff
@@ -0,0 +1,78 @@
+
+        # concatenate basins
+        basins_all = pd.concat([AP, WP, IO, SH])
+        basins_all.to_csv(f'{output}rhab_exposure/wp{pop_year}_rhab{hab_year}.csv')
```
c.f. comment above.

chahank commented:
Please add also a readme, for instance a markdown file explaining the script and how to use it and the link to the publication (when it will be available).

sarah-hlsn committed:
Update exposure_ecosystem_matching.py

added copyright statement and changed file paths

sarah-hlsn committed:
Update exposure_ecosystem_matching.py

added explanatory comments

chahank commented on 202305_coastal_ecosystems_TC/impact_calculation/exposure_ecosystem_matching.py:
```diff
@@ -0,0 +1,101 @@
+pop_years = ['2000', '2020']
+hab_years = ['1992', '2020']
+basins = ['AP', 'WP', 'IO', 'SH']
+input = '/Users/sarah/Documents/CLIMADA/coastal_ecosystems/data/'   # path to input data 
+output = '/Users/sarah/Documents/CLIMADA/coastal_ecosystems/results/intermediate/'  # path to save results
```
You do not want to have your personal links on this file. Please make a generic file path and explain in the readme where to place the data (and how to get it).

sarah-hlsn committed:
Update exposure_ecosystem_matching.py

generic file path

sarah-hlsn committed:
Update exposure_ecosystem_matching.py

added missing quotation marks

sarah-hlsn committed:
Add files via upload

modules needed to create impact functions and calculate impact per windspeed category

chahank commented on 202305_coastal_ecosystems_TC/impact_calculation/exposure_ecosystem_matching.py:
```diff
@@ -0,0 +1,101 @@
+pop_years = ['2000', '2020']
+hab_years = ['1992', '2020']
+basins = ['AP', 'WP', 'IO', 'SH']
+input_path = '/Users/USERNAME/Documents/CLIMADA/coastal_ecosystems/data/'   # path to input data 
+output_path = '/Users/USERNAME/Documents/CLIMADA/coastal_ecosystems/results/intermediate/'  # path to save results
```
Why put such a Path? This is specific to the MacOS. The trick is to use relative paths.

```suggestion
input_path = '../data/'   # path to input data 
output_path = '../'  # path to save results
```

Then add a folder `impact_calculation/data/`.

chahank replied:
Also, in if you want to make it very simple for cross-platform users, you can use the Pathlib library instead of using strings to define paths.

sarah-hlsn committed:
Update 202305_coastal_ecosystems_TC/impact_calculation/exposure_ecosystem_matching.py

Co-authored-by: Chahan M. Kropf <chahan.kropf@usys.ethz.ch>

chahank commented:
Great! Now please just add the correct folder structure for the data and a brief readme file to explain what the code does and how to use it. Once the article is published, we shall also add the doi reference.

sarah-hlsn committed:
Update exposure_ecosystem_matching.py

changed copyright statement

sarah-hlsn committed:
Update CatImpactFuncTC.py

added author

sarah-hlsn committed:
Update CatImpactTC.py

added author

sarah-hlsn committed:
Add files via upload

script to calculate impacts

sarah-hlsn committed:
Update exposure_ecosystem_matching.py

changed output file path to match path in impact calculation file

sarah-hlsn committed:
Create global_analysis.py

script to analyse global impact data

sarah-hlsn committed:
Add files via upload

script for regional and country-level analysis

sarah-hlsn committed:
Add files via upload

added read_me file

chahank commented on 202305_coastal_ecosystems_TC/README.md:
```diff
@@ -0,0 +1,22 @@
+| Population exposure                    | Global gridded population data, 1km resolution   | WorldPop, 2018. Global 1km Population. https://doi.org/10.5258/SOTON/WP00647                                                                                                                                                                                                                                      ...
+| Coastal ecosystem protective rank data | Shoreline points, 450m resolution                | Natural Capital Project, 2019. InVEST [WWW Document]. Nat. Cap. Proj. URL https://naturalcapitalproject.stanford.edu/software/invest                                                                                                                                                                              ...
+| STORM TC hazard data                   | Synthetic tropical cyclone tracks                | Bloemendaal, N., Haigh, I.D., de Moel, H., Muis, S., Haarsma, R.J., Aerts, J.C.J.H., 2020. Generation of a global synthetic tropical cyclone hazard dataset using STORM. Sci. Data 7, 40. https://doi.org/10.1038/s41597-020-0381-2                                                                               ...
+| STORM-C TC hazard data                 | Synthetic tropical cyclone tracks SSP585 in 2050 | Bloemendaal, N., de Moel, H., Martinez, A.B., Muis, S., Haigh, I.D., van der Wiel, K., Haarsma, R.J., Ward, P.J., Roberts, M.J., Dullaart, J.C.M., Aerts, J.C.J.H., 2022. A globally consistent local-scale assessment of future tropical cyclone risk. Sci. Adv. 8, eabm8438. https://doi.org/10.1126/sciadv.abm ...
```
```suggestion
| Data type                              | Description                                      | Data source                                                                                                                                                                                                                                                                                                           | Data preparation                                                                     | 
|----------------------------------------|--------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------|
| Landcover data                         | Global land cover satellite data                 | SA CCI-LC, 2017. ESA/CCI viewer [WWW Document]. URL http://maps.elie.ucl.ac.be/CCI/viewer/download.php                                                                                                                                                                                                                | Compute coastal ecosystem protective rank using InVEST Coastal Vulnerability model   |
| Population exposure                    | Global gridded population data, 1km resolution   | WorldPop, 2018. Global 1km Population. https://doi.org/10.5258/SOTON/WP00647                                                                                                                                                                                                                                          | Clip to 10m elevation above sea level using DEM, Vectorize, Clip to TC basin extents | 
| Coastal ecosystem protective rank data | Shoreline points, 450m resolution                | Natural Capital Project, 2019. InVEST [WWW Document]. Nat. Cap. Proj. URL https://naturalcapitalproject.stanford.edu/software/invest                                                                                                                                                                                  | Clip to TC basin extents                                                             | 
| STORM TC hazard data                   | Synthetic tropical cyclone tracks                | Bloemendaal, N., Haigh, I.D., de Moel, H., Muis, S., Haarsma, R.J., Aerts, J.C.J.H., 2020. Generation of a global synthetic tropical cyclone hazard dataset using STORM. Sci. Data 7, 40. https://doi.org/10.1038/s41597-020-0381-2                                                                                   | Compute windfields using CLIMADA Hazard module, with 300 arc seconds resolution      |
| STORM-C TC hazard data                 | Synthetic tropical cyclone tracks SSP585 in 2050 | Bloemendaal, N., de Moel, H., Martinez, A.B., Muis, S., Haigh, I.D., van der Wiel, K., Haarsma, R.J., Ward, P.J., Roberts, M.J., Dullaart, J.C.M., Aerts, J.C.J.H., 2022. A globally consistent local-scale assessment of future tropical cyclone risk. Sci. Adv. 8, eabm8438. https://doi.org/10.1126/sciadv.abm8438 | Compute windfields using CLIMADA Hazard module, with 300 arc seconds resolution      | 
```

chahank replied:
I think you do not need the extra columns right?

chahank commented on 202305_coastal_ecosystems_TC/impact_calculation/Impact_calculation.py:
```diff
@@ -0,0 +1,33 @@
+Author: Sarah Hülsen
+"""
+
+from climada.entity import Exposures
+from climada.hazard import Hazard
+from modules.CatImpactFuncTC import cat_impf_set
+from modules.CatImpactTC import imp_per_cat
+
+
+basin = 'AP'    # TC basin (e.g. AP, IO, SH, WP)
+year = '2020'   # population exposure year (e.g. 2000 or 2020)
+scenario = 'hist_STORM'  # current climate or SSP585 TC hazard (hist_STORM or clim_STORM_{model})
+
+# Data paths
+input_path = '../results/intermediate/'
+output_path = '../results/final/'
```
```suggestion
from pathlib import Path

from climada.entity import Exposures
from climada.hazard import Hazard
from modules.CatImpactFuncTC import cat_impf_set
from modules.CatImpactTC import imp_per_cat


basin = 'AP'    # TC basin (e.g. AP, IO, SH, WP)
year = '2020'   # population exposure year (e.g. 2000 or 2020)
scenario = 'hist_STORM'  # current climate or SSP585 TC hazard (hist_STORM or clim_STORM_{model})

# Data paths
input_path = Path('../results/intermediate/')
output_path = Path('../results/final/')
```

It is convenient to use the `pathlib` library as this takes care of the different paths between operating systems. Your path here would not work on windows for example.

I suggest you edit all path definitions using this library.

chahank reviewed (commented):
Excellent! Two small suggestions. Also, I would add the DOI of the paper if there is a preprint already available.

sarah-hlsn committed:
Update 202305_coastal_ecosystems_TC/README.md

Co-authored-by: Chahan M. Kropf <chahan.kropf@posteo.de>

sarah-hlsn committed:
Update 202305_coastal_ecosystems_TC/impact_calculation/Impact_calculation.py

Co-authored-by: Chahan M. Kropf <chahan.kropf@posteo.de>

sarah-hlsn committed:
Update exposure_ecosystem_matching.py

added pathlib

sarah-hlsn committed:
Update global_analysis.py

added pathlib

sarah-hlsn committed:
Update per_country_analysis.py

added pathlib

sarah-hlsn committed:
Update region_matching.py

added pathlib

sarah-hlsn committed:
Update regional_analysis.py

added pathlib

sarah-hlsn committed:
Update README.md

added doi link to preprint
